Connect HTML spec up to new CSS `interactivity` property

mfreed7 opened the pull request:
I was a bit unsure about a few parts of this PR, so help would be appreciated from the editors. The idea is:
1. Remove duplication - allow the CSS spec to define the *concept* of "inert".
2. Make the `inert` HTML attribute set `interactivity:inert` via a new UA stylesheet rule.
3. Make modal dialogs non-inert via an `interactivity:auto` UA stylesheet rule.
4. {**later**} A rule likely needs to be added for fullscreen elements also, setting `:fullscreen {interactivity: auto}`.

Side-note: at this time the CSS spec isn't being published. Bot problems. But [this CSS spec PR](https://github.com/w3c/csswg-drafts/pull/11178/files) landed and should eventually contain the references added in this PR.

-----

- [ ] At least two implementers are interested (and none opposed):
   * …
   * …
- [ ] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * … <!-- If these tests are tentative, link a PR to make them non-tentative. -->
- [ ] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: …
   * Gecko: …
   * WebKit: …
   * Deno (only for timers, structured clone, base64 utils, channel messaging, module resolution, web workers, and web storage): …
   * Node.js (only for timers, structured clone, base64 utils, channel messaging, and module resolution): …
- [ ] Corresponding [HTML AAM](https://w3c.github.io/html-aam/) & [ARIA in HTML](https://w3c.github.io/html-aria/) issues & PRs:
- [ ] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: …
- [ ] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

lilles commented on source:
```diff
@@ -79283,39 +79287,32 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
   that is <span>connected</span> to <var>document</var>, with the exception of the
   <var>subject</var> element and its <span>flat tree</span> descendants, must become
-  <span>inert</span>.</p>
+  <span>inert</span>. This is achieved via the following UA stylesheet rule:</p>
```
I don't this is accurate. The host language inertness does not affect the interactivity property and the interactivity property cannot be used to escape host language inertness, so I think the effect of this rule is to escape interactivity property inertness from ancestor elements of the modal dialog.

For instance, if the two dialogs below are modal, they will not be inert when rendered in the top layer:

```html
<div inert>
  <dialog></dialog>
</div>
<div style="interactivity:inert">
  <dialog></dialog>
</div>
```

lilles replied:
lgtm

lilles commented on source:
```diff
@@ -79332,7 +79329,11 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
   <p>The <code data-x="attr-inert">inert</code> attribute is a <span>boolean attribute</span> that
   indicates, by its presence, that the element and all its <span>flat tree</span> descendants which
   don't otherwise escape inertness (such as modal dialogs) are to be made <span>inert</span> by the
```
This text still looks like it's applying host language inertness that cannot be overridden by author styles. I think we should rephrase this somewhat.

mfreed7 replied:
Yep, you're right. I removed most of this, and moved around the UA stylesheet to be right here.

lilles replied:
> Yep, you're right. I removed most of this, and moved around the UA stylesheet to be right here.

lgtm

lilles commented on source:
```diff
@@ -79283,39 +79287,32 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
+
+  <pre><code class="css">dialog:modal {
+  interactivity: auto;
+}</code></pre>
```
We should add a UA rule for `:fullscreen` in the appropriate location as well.

lilles replied:
Probably a separate PR for https://fullscreen.spec.whatwg.org/

mfreed7 replied:
I think I agree - I just wanted to get confirmation from an editor. I just don't see any other CSS rules listed there, and I never like to be the first one to add something new to a spec.

nt1m replied:
@mfreed7 @lilles No need to make a separate PR for fullscreen. `:modal { interactivity: auto }` covers both modal dialogs and fullscreen :)

lilles replied:
Hm. I didn't know.

Working on the impl, the order/specificity of `[inert]` and `:modal` matters and is covered by existing tests in for instance: https://wpt.live/inert/inert-with-modal-dialog-003.html

If a modal dialog has an inert attribute, it's expected that it wins, so `[inert]` needs to appear after `:modal` in the UA sheet.

mfreed7 replied:
I've added both rules, together, in the Rendering section.

mfreed7 reviewed (commented):
Thanks for the review!

nt1m commented:
I think `blocked by a modal dialog` should be defined in terms of `interactivity` (either set as a presentational hint on the root, or using an UA stylesheet rule), because otherwise `dialog:modal { interactivity: auto }` won't do anything (because `auto` is the default value. I don't think this is what the current HTML spec PR does.

mfreed7 commented:
> I think `blocked by a modal dialog` should be defined in terms of `interactivity` (either set as a presentational hint on the root, or using an UA stylesheet rule), because otherwise `dialog:modal { interactivity: auto }` won't do anything (because `auto` is the default value. I don't think this is what the current HTML spec PR does.

I added a `Note` about this. But I think the concept (agreed to over in CSSWG I believe) was that both specs can set something to be inert. The problem with defining modal dialog behavior in terms of CSS is that it allows author CSS to make parts of the page "un-inerted", which is undesirable. And it can't be set to `interactivity:inert/auto !important` to avoid that problem, because then that doesn't allow the author to "re-inert" the dialog itself.

lilles commented:
> > I think `blocked by a modal dialog` should be defined in terms of `interactivity` (either set as a presentational hint on the root, or using an UA stylesheet rule), because otherwise `dialog:modal { interactivity: auto }` won't do anything (because `auto` is the default value. I don't think this is what the current HTML spec PR does.
> 
> I added a `Note` about this. But I think the concept (agreed to over in CSSWG I believe) was that both specs can set something to be inert. The problem with defining modal dialog behavior in terms of CSS is that it allows author CSS to make parts of the page "un-inerted", which is undesirable. And it can't be set to `interactivity:inert/auto !important` to avoid that problem, because then that doesn't allow the author to "re-inert" the dialog itself.

Another issue is that no one's been able to come up with UA styles that actually works with shadow DOM.

noamr commented on source:
```diff
@@ -79283,56 +79287,87 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
+}</code></pre>
+
+  <div class="note">
+   <p>Since the <code data-x="attr-inert">inert</code> attribute simply adds
```
We don't "simply" in the HTML standard :)

mfreed7 replied:
Yeah, that's definitely true. Done.

noamr reviewed (commented):
I've stumbled upon this type of problem when implementing `moveBefore` things... Thanks for working on this! Added some non-authoritative nits

domenic commented on source:
```diff
@@ -79283,56 +79287,87 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
 
-  <p>A node (in particular elements and text nodes) can be <dfn>inert</dfn>. When a node is
-  <span>inert</span>:</p>
+  <p>A node (in particular elements and text nodes) can be <dfn>inert</dfn>.</p>
```
It feels like there are now three concepts:

- A node is "marked as inert": done by specific APIs like dialog modal blocking
- A node "is inert": true if either interactivity or "marked as inert"
- An element or text node "behaves inertly": defined by CSS
  - But in a somewhat confusing way, since find-in-page and editability are defined by HTML, and some aspects of text selection (like the JS methods that cause it) are as well?
  - But I guess this was a preexisting problem; the paragraph you're deleting was already an action-at-a-distance part of the spec

I'm unsure whether it's worth separating these three concepts explicitly, but I wanted to call this out as potentially tricky.

I'm particularly confused what it means for non-element/text nodes to be inert now.

mfreed7 replied:
Thanks for raising this. I did have some trouble with this PR trying to decide what all of this should be called, or where it should live.

One nit: a node that "is inert" always "behaves inertly" so that's kind of one concept, right?

Anyway, happy to modify this however you see fit. Pointers appreciated. In particular the find-in-page, text selection, etc. are all (now) [defined in CSS](https://drafts.csswg.org/css-ui-4/#inertness), so I at least wanted to have just one place where that's done.

domenic replied:
> One nit: a node that "is inert" always "behaves inertly" so that's kind of one concept, right?

Right, except that "behaves inertly" only applies to text and element nodes, and the definition lives in a different document from "is inert".

My suggestion, although I don't feel too strongly:

- Have only one definition of "inert", the one in CSS. I guess it only applies to elements and text nodes.
- Define "marked as inert" or maybe "marked as explicitly inert" in HTML, as a boolean which all nodes have. Update that boolean when blocked by a modal dialog.
- Have a sentence like "An element or text node must be considered [inert] if it is marked as inert or...".
  - This links to the CSS definition instead of having its own.
  - This mechanism of having CSS define the concept and HTML say when it applies is similar to what we do for https://html.spec.whatwg.org/#pseudo-classes

domenic replied:
Hmm yeah, I see what you mean. It doesn't seem worth it to transform this into an explicit boolean on every node, as maintaining it would be a lot of work.

A middle path that stays close to the current spec while also clarifying the interaction between CSS and blocked-by-a-modal would be something like:

- Modify the current "blocked by a modal dialog" to something like:

    > A Document _document_ is **blocked by a modal dialog** _subject_ if _subject_ is the topmost dialog element in document's top layer. While _document_ is so blocked, every <del>node</del><ins>element or text node</ins> that is connected to _document_, with the exception of the _subject_ element and its flat tree descendants, <del>must become inert</del><ins>is considered **inert via dialog blockage**.
  
  (or maybe "explicitly inert" or something like that.)

- Remove the explicit "Let subject's node document be blocked by the modal dialog subject." As you say, that seems redundant and broken. Probably we forgot to remove it when we improved top layer + dialog interactions in the past.

    We can relocate the note about "This will cause the..." to the next step about top layer.

- Now we can go with something like my suggestion above, of having "An element or text node must be considered [inert] if it is [inert via dialog blockage] or (the CSS stuff)".

But I kind of like the direction you're currently going as well, as it's quite explicit.

I wonder if there's any inspiration we can take from implementations here? How is it implemented in Chromium? (And in particular, how is the interaction between the CSS interactivity and the blocked-by-a-modal-dialog interactivity handled.)

The reason I ask about how the implementation works is that I'm worried about leading us toward a spec that _seems_ explicit and rigorous, but isn't actually easy to implement / close to existing implementations.

mfreed7 replied:
@lilles can you comment on this? I agree that it seems best to try to follow an implementation, to make sure it makes sense.

lilles replied:
> I wonder if there's any inspiration we can take from implementations here? How is it implemented in Chromium? (And in particular, how is the interaction between the CSS interactivity and the blocked-by-a-modal-dialog interactivity handled.)

It is implemented by setting a flag on the computed style for inert-via-dialog-or-fullscreen that we name IsHTMLinert, which is the host language inertness that the spec for `interactivity` talks about:

https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/css/resolver/style_adjuster.cc;l=948-987

The computed value for `interactivity` is a separate value on the computed styles.

We decide if inert-effects happen based on or'ing these.

mfreed7 replied:
Great, thanks for the feedback. So @domenic, it sounds like the implementation actually roughly mirrors [your suggestion](https://github.com/whatwg/html/pull/10956/files#r1948443606). Based on that, I've reverted a bit of my prior commit, and went in the direction you recommended. I used "blocked by a modal" (remove "dialog") and "inert via modal blockage", so that we can connect up fullscreen and it won't be odd. Otherwise, I think it looks ok. PTAL!

domenic commented on source:
```diff
@@ -79283,56 +79287,87 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
+  data-x="css-inert">inert</span>. Inert nodes generally cannot be focused, and user agents do not
+  expose the inert nodes to accessibility APIs or assistive technologies. Inert nodes that are <span
   data-x="concept-command">commands</span> will become inoperable to users, in the manner described
   above.</p>
```
The manner is no longer described above.

mfreed7 replied:
Well, there's still this, from around line 61426:

> `<p>`The `<span data-x="command-facet-DisabledState">`Disabled State`</span>` facet of the command is
  true if the element or one of its ancestors is `<span>`inert`</span>`, and false otherwise.`</p>`

...which I didn't touch. That's the only place this is defined, since the CSS spec doesn't talk about commands.

domenic replied:
Hmm, fair enough.

(The whole "command" section is historical, and probably overengineered and somewhat fictitious. It seems to come from the idea that browsers might provide a sort of command palette UI for exposing all the commands on a page. So we keep just making minor updates around its edges. Oh well.)

domenic commented on source:
```diff
@@ -79283,56 +79287,87 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
   data-x="add an element to the top layer">adding</span> the <code>dialog</code> element to its
   <span>node document</span>'s <span>top layer</span>.</p>
 
+  <p>The UA will add the following stylesheet rule:</p>
```
This belongs in https://html.spec.whatwg.org/#rendering, not here. That can also help with the specificity problem mentioned elsewhere by collocating with the inert attribute.

mfreed7 replied:
Ok, moved. I inserted it into the [Flow Content section](https://html.spec.whatwg.org/#flow-content-3), where it seemed to mesh nicely with existing modal dialog stuff. LMK if that makes sense.

domenic commented on source:
```diff
@@ -79283,56 +79287,87 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
+  sets the <span data-x="css-interactivity">'interactivity'</span> property to <span
+  data-x="interactivity-inert">'inert'</span> via a UA stylesheet rule:</p>
+
+  <pre><code class="css">@namespace "http://www.w3.org/1999/xhtml";
```
This also belongs in the rendering section.

mfreed7 replied:
Done.

domenic commented on source:
```diff
@@ -79283,56 +79287,87 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
-  don't otherwise escape inertness (such as modal dialogs) are to be made <span>inert</span> by the
-  user agent.</p>
+  sets the <span data-x="css-interactivity">'interactivity'</span> property to <span
+  data-x="interactivity-inert">'inert'</span> via a UA stylesheet rule:</p>
```
Attribute descriptions should be more functional (or semantic, if possible). The interaction with CSS is confined to the rendering section.

The hidden attribute is a good example to follow.

mfreed7 replied:
Thanks for the pointer to the hidden attribute - that was helpful. I think I've modified this accordingly, but let me know.

mfreed7 commented on source:
```diff
@@ -79283,56 +79287,87 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
+}</code></pre>
+
+  <div class="note">
+   <p>The HTML spec defines several ways for an element to be <span>inert</span>,
```
Done, x3.

mfreed7 commented on source:
```diff
@@ -79283,56 +79287,87 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
+  <div class="note">
+   <p>The HTML spec defines several ways for an element to be <span>inert</span>,
+   including the <code data-x="attr-inert">inert</code> attribute, and the <span>blocked by a modal
+   dialog</span> concept. The CSS spec additionally provides the <span
```
Done, I think? I'm not sure I linked that in the approved way - let me know.

domenic replied:
HTML is a bit silly about this. The usual method is `The <cite>CSS Basic User Interface Module</cite> specification ... rest of sentence. <ref>CSSUI</ref>` With the reference at the end of the paragraph.

mfreed7 commented on source:
```diff
@@ -79283,56 +79287,87 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
+   including the <code data-x="attr-inert">inert</code> attribute, and the <span>blocked by a modal
+   dialog</span> concept. The CSS spec additionally provides the <span
+   data-x="css-interactivity">'interactivity'</span> property. These methods don't directly affect
+   each other, but do interact with each other, which is why several UA stylesheet rules are added
```
Better, thanks.

mfreed7 reviewed (commented):
Thanks for the comments!

domenic commented on source:
```diff
@@ -79283,56 +79287,80 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
   <span>node document</span>'s <span>top layer</span>.</p>
 
+  <div class="note">
+   <p>The HTML specification defines several ways for an element to be <span>inert</span>,
```
```suggestion
   <p>The HTML Standard defines several ways for an element to be <span>inert</span>,
```

domenic commented on source:
```diff
@@ -135626,6 +135654,14 @@ dialog:popover-open {
   display:block;
 }
 
+:modal {
```
Hmm, honestly all of :modal, [inert], and the existing popover stuff are not really about "flow content". They would probably benefit from a new section like "Hidden elements" has currently.

But, I don't think it's really necessary to fix that in this PR, so just using the "Flow content" section for now, and maybe filing a follow-up bug, would be good.

(I'm also a bit uneasy about how whenever we split up the UA stylesheet, we make the order of declarations less clear, which might matter in some cases. So a separate issue might give us some space to discuss that as well.)

domenic commented on source:
```diff
@@ -972,7 +972,7 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
   <!-- NON-NORMATIVE SECTION -->
 
-  <p>Unlike previous versions of the HTML specification, this specification defines in some detail
+  <p>Unlike previous versions of the HTML Standard, this specification defines in some detail
```
Technically previous versions weren't the HTML Standard, so this instance should remain unchanged. (They were instead, e.g., the HTML 2.0 W3C Recommendation, or similar.)

mfreed7 replied:
Done. (And confusing! 😆 )

domenic commented on source:
```diff
@@ -79283,56 +79284,81 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
-  <p>A node (in particular elements and text nodes) can be <dfn>inert</dfn>. When a node is
-  <span>inert</span>:</p>
-
+  <p>An element or text node <var>node</var> must be considered <span>inert</span> if one of the
```
```suggestion
  <p>An element or <code>Text</code> node <var>node</var> must be considered <span>inert</span> if one of the
```

mfreed7 replied:
Done.

domenic reviewed (approved):
LGTM with nits. This turned out pretty nice; thanks for iterating!

Let me know when the checkboxes in the OP are filled out.

mfreed7 commented:
> LGTM with nits. This turned out pretty nice; thanks for iterating!
> 
> Let me know when the checkboxes in the OP are filled out.

Thanks for the review!

@annevk @emilio would you mind please providing support for landing this?

mfreed7 commented:
> * [ ]  Corresponding [HTML AAM](https://w3c.github.io/html-aam/) & [ARIA in HTML](https://w3c.github.io/html-aria/) issues & PRs:

@scottaohara @aleventhal does this need any change to the HTML AAM spec? This PR (to HTML) doesn't actually change the behavior of existing things like `inert` or modal dialogs. The CSSWG spec change (which already landed, [here](https://drafts.csswg.org/css-ui-4/#inertness)) might need something?

annevk commented:
What is the status of this in the CSS WG? It's relatively straightforward to get things in drafts, but that historically hasn't really conveyed any kind of stability.

emilio commented on source:
```diff
@@ -135626,6 +135652,14 @@ dialog:popover-open {
   display:block;
 }
 
+:modal {
```
Is this correct? When you have multiple modal elements in the top layer, don't all match `:modal`? You need a magic "topmost modal" thing here, right?

mfreed7 replied:
No, I don't think so. Yes, this sets the `interactivity` to `auto` for all dialogs, but see the `<dfn>blocked by a modal</dfn>` section. In particular, there are some `Note`s about this behavior there. That section blocks all but the topmost via "html inertness". This CSS rule is here just to allow authors to re-override `interactivity` back to `inert` if they so desire.

chrishtr commented:
> What is the status of this in the CSS WG?

In terms of the CSSWG's process (which follows W3C practices), it is in an editor's draft. There have been multiple CSSWG consensus resolutions, such as to add it and decide on behaviors like interaction with the top layer.

astearns commented:
> > What is the status of this in the CSS WG?
> 
> In terms of the CSSWG's process (which follows W3C practices), it is in an editor's draft. There have been multiple CSSWG consensus resolutions, such as to add it and decide on behaviors like interaction with the top layer.

But (IMO) we should not consider anything in CSSWG drafts as stable until there is at least a prototype and some tests. Lack of unresolved issues before that happens does not mean much

mfreed7 commented:
> But (IMO) we should not consider anything in CSSWG drafts as stable until there is at least a prototype and some tests. Lack of unresolved issues before that happens does not mean much

Just to put it here, Blink has implemented this property, and written the WPTs:

https://wpt.fyi/results/css/css-ui?label=master&label=experimental&aligned&q=interactivity

emilio commented:
Given what we discussed today this looks good to me, but we should make very clear that the "make modal dialogs not overridable, but inert attribute overridable" is intentional.

mfreed7 commented on source:
```diff
@@ -79283,56 +79284,89 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
+  &lt;dialog>&lt;/dialog>
+&lt;/div></code></pre>
+
+    <p>The reason that HTML "overrides" the <span data-x="css-interactivity">'interactivity'</span>
```
@emilio I added this chunk as a result of today's discussion. LMK if that helps or hurts.

mfreed7 commented:
> Given what we discussed today this looks good to me, but we should make very clear that the "make modal dialogs not overridable, but inert attribute overridable" is intentional.

Great, thanks! I've checked the checkbox in the OP based on this comment.

I also added a paragraph of more motivation for the "weird" behavior. LMK what you think.

domenic commented on source:
```diff
@@ -79283,56 +79284,89 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
+
+    <p>The reason that HTML "overrides" the <span data-x="css-interactivity">'interactivity'</span>
+    property in the case of modal dialogs is that accessibility technology requires modal dialogs to
+    be truly modal. Content outside of an open modal dialog should always be inert. For this reason,
```
Nit: should -> "needs to be" or similar to avoid magic normative words in non-normative notes.

domenic commented:
All the boxes here looked checked, but agreement is fresh off the presses and maybe people want to review the new note.

We haven't gotten an answer back on whether this affects HTML-AAM / ARIA, but I am pretty sure it doesn't since there's no new HTML attributes or elements involved.

I'll plan on merging this mid-next-week, assuming @mfreed7 addresses my nit and rebases on master to resolve conflicts by then.

annevk commented:
No, this cannot merge. The CSS side still has open issues and the CSS WG has not indicated this is stable to reference.

domenic commented:
We don't have any issues in our working mode for "the CSS side has open issues" or blocking on W3C process. I believe we meet all the criteria in https://whatwg.org/working-mode#additions and https://whatwg.org/working-mode#changes . Can you point to which one is not met?

emilio commented:
https://github.com/w3c/csswg-drafts/issues/11849 is Luke's issue (for cross-referencing purposes)

mfreed7 commented:
> We haven't gotten an answer back on whether this affects HTML-AAM / ARIA, but I am pretty sure it doesn't since there's no new HTML attributes or elements involved.

I'm pretty sure too, but @scottaohara did mention to me yesterday that he'd take a look at this soon. I'm ok waiting for that, just to confirm.

> I'll plan on merging this mid-next-week, assuming @mfreed7 addresses my nit and rebases on master to resolve conflicts by then.

Ok, I addressed the nit, and rebased. The rebase wasn't super clean, given the dialog closedby changes, but I think I got it fixed up correctly. An extra pair of eyes wouldn't hurt.

alice commented:
This is really disappointing to see. I thought the discussion on https://github.com/whatwg/html/issues/10811 was really promising.

mfreed7 commented:
> This is really disappointing to see. I thought the discussion on #10811 was really promising.

Sorry to hear that. The discussion on #10811 was hypothetical and (I think?) could still be moved forward if we wanted to. It describes a way to have explicit exceptions for other top layer things like popovers while there's a modal dialog open. My understanding from several comments on that issue was that this situation wasn't desirable from an a11y perspective, so this PR didn't change the modal dialog behavior. Can you comment? E.g. [this comment](https://github.com/whatwg/html/issues/10811#issuecomment-2540307654) mentions that this might violate screen reader assumptions, and the comments about re-parenting things in the a11y tree scared me a bit. Perhaps those aren't as scary as they sound?

scottaohara commented:
from what i can tell, reading this PR over a few times now, is that i don't see anything in here that would necessarily need to be changed in the HTML AAM or CORE AAM specs - unless there are mappings that were altered that wouldn't be called out in this HTML spec update?

As I mentioned in my initial comment about the use cases i had for allowing something to escape inertness - I wasn't wedded to the idea that it needed to be done via CSS.  I'm not against it, but I'm also not a strong supporter of it being the solution here. 

The back and forth with Alice in #10811 seemed promising - specifically allowing popovers to escape inertness, and then essentially adjusting them in the a11y tree - much like aria-owns does - to make it seem to people using AT that the popover was a child of the current modal dialog. 

So, I guess with all that said, I'm not sure about allowing for inertness to be escaped if we already have one potential solution which would arguably be a lot safer / better for the end user experience.  For my other use cases, while I still think they're valid, I think it's probably a better idea to see if there aren't other potential solutions, than introducing an inert escape hatch that can never be closed once open.  Because use cases aside, @alice is right that this could be problematic if unchecked. If this PR can go through without the ability to uninert things for now, that'd be the safer option, imo.

alice commented:
> It describes a way to have explicit exceptions for other top layer things like popovers while there's a modal dialog open.

That's not how I had envisioned it. The way I proposed and understood it (and it seemed like you [agreed](https://github.com/whatwg/html/issues/10811#issuecomment-2555956040)) was that it would both _explain_ how dialogs and fullscreen elements made the rest of the page inert, and fix the issue with content in a top layer on top of dialogs being made inert erroneously.

It would also open the door to exploring better ways to implement custom modal UI without the need to use `<dialog>` _or_ `inert` (and thus with no need to add the capacity to escape `inert` here).

> E.g. [this comment](https://github.com/whatwg/html/issues/10811#issuecomment-2540307654) mentions that this might violate screen reader assumptions, and the comments about re-parenting things in the a11y tree scared me a bit. Perhaps those aren't as scary as they sound?

Those are implementation details needed to make the a11y tree semantics match the dialog semantics, and would be needed regardless (i.e. if the browser didn't do this, the author using `popover` with either a `<dialog>` or any custom dialog which used `role=dialog` would have to, and they most likely wouldn't know to).

mfreed7 commented:
Hi, I’ve tried below to answer all questions. Because there were many, this got quite long. Apologies for length or if I missed anything!

> from what i can tell, reading this PR over a few times now, is that i don't see anything in here that would necessarily need to be changed in the HTML AAM or CORE AAM specs - unless there are mappings that were altered that wouldn't be called out in this HTML spec update?

Great! Thanks for that - based on this comment I’ve checked the last box on this PR.

> As I mentioned in my initial comment about the use cases i had for allowing something to escape inertness - I wasn't wedded to the idea that it needed to be done via CSS. I'm not against it, but I'm also not a strong supporter of it being the solution here.

I agree there are use cases - e.g. a “user space” modal dialog that doesn’t utilize the top layer. Currently, that’s typically implemented using manual focus trapping, which is worse for a11y and correctness.

> The back and forth with Alice in #10811 seemed promising - specifically allowing popovers to escape inertness, and then essentially adjusting them in the a11y tree - much like aria-owns does - to make it seem to people using AT that the popover was a child of the current modal dialog.

My concerns are twofold: First, I wonder if the better “solution” to that problem is to require developers to simply nest the popover within the dialog. That provides out-of-the-box inertness and a11y. Second, I worry about the implementability of making the a11y tree not match the DOM tree in such a significant way. I know that Blink has had a long series of bugs related to situations like this, and we’d like to avoid more of them. + @aleventhal for comments based on experience here.

>  I think it's probably a better idea to see if there aren't other potential solutions, than introducing an inert escape hatch that can never be closed once open. 

Are there other solutions that are better? It seems that the crux of the issue isn’t really the shape of *how* something gets un-inerted, but more that it can be un-inerted at all.

> If this PR can go through without the ability to un-inert things for now, that'd be the safer option, imo.

Technically, this PR doesn’t add un-inerting behavior (that behavior is in the [CSS spec](https://drafts.csswg.org/css-ui/#inertness)). On the contrary, this PR goes to some length to preserve the **existing** behavior wherein modal dialog related inertness **cannot** be un-inerted.

> That's not how I had envisioned it. The way I proposed and understood it (and it seemed like you [agreed](https://github.com/whatwg/html/issues/10811#issuecomment-2555956040)) was that it would both _explain_ how dialogs and fullscreen elements made the rest of the page inert, and fix the issue with content in a top layer on top of dialogs being made inert erroneously.

> It would also open the door to exploring better ways to implement custom modal UI without the need to use `<dialog>` _or_ `inert` (and thus with no need to add the capacity to escape `inert` here).

Can you help me understand your approach a bit better? Reading [your comment](https://github.com/whatwg/html/issues/10811#issuecomment-2540247720) and the ones below, it seems to just say that modal dialogs form a “blocking” top layer item, which un-inert the things above them in the stack, and inert everything below them. That still feels like magic (i.e. it doesn’t _explain_ in the sense of other developer-visible APIs) and also *requires* the use of the dialog element and the UA-provided top layer. I.e. it doesn’t allow userspace modals that aren’t in the top layer. Please help me understand my disconnect.

> Those are implementation details needed to make the a11y tree semantics match the dialog semantics, and would be needed regardless (i.e. if the browser didn't do this, the author using `popover` with either a `<dialog>` or any custom dialog which used `role=dialog` would have to, and they most likely wouldn't know to).

One nice thing about the existing behavior of popovers and modal dialogs is that it in fact does require the developer to properly nest the popover within the dialog, or it’ll be inert and won’t work. I’m concerned about making the implementation of the a11y more convoluted, for fear of introducing major bugs and crashes and things, which aren’t good for users.

> i'm currently looking at the mdn draft for the customizable select, and in it a passing comment was mentioned that the inert button part of the select could have its inertness undone using the CSS interactivity property. That button part was made inert on purpose. The HTML AAM mappings specifically call it out as inert and that as far as the a11y tree is concerned, the select/button part should not be perceived as two separate elements to end users - why would authors be allowed to undo that with CSS?

This was simply a general effort to avoid `!important` on the UA stylesheet for customizable-`<select>`. There might be uncommon use cases where the button wants to be made back into a button, hopefully with careful consideration of a11y and use of ARIA. But I would consider this a fairly corner case, and not likely to be hit by developers on accident or without noticing.

> I think the idea of a CSS opt-out of inert makes me nervous, especially one that differs in behavior depending on what has made the element is inert (i.e. because of an ancestor style vs. a modal dialog). In practice, in a dev environment with code contributions from many different teams (as in any medium-to-large organization), it's easy to have controls from different sources using different methods to inert background content. One dialog might easily be an HTML `<dialog>`, and another a custom dialog component using `inert` or the new CSS style. Or, as in the case of the UI library I work on, we might do things like convert our shipped Dialog control from a custom solution to one using `<dialog>` in a version update.

So I think one nice thing about the current PR is that the mechanism for un-inerting is always the same: `interactivity: auto`. That’s true whether the inertness was provided by the `inert` attribute or `interactivity: inert`. Of course, the one exception is modal `<dialog>`s, and that exception was preserved precisely and only to maintain the a11y guarantees that it already provides.


> In that environment, if another control (e.g. a notification popup) sets the `interactivity` style, that would be exceedingly frustrating to debug and maintain since it might work with some dialogs but not the others. Conversely, a control that incorrectly sets `interactivity: auto` would still work fine with a `<dialog>` element but not the custom control. This difference in behavior would, in my opnion, make the style effectively verboten since you (a theoretical component dev) can't control or predict how it'll behave re: other inerting controls on the page that you have no control over. I also don't really think it's reasonable to expect regular non-standards-participating developers to understand the distinction between HTML dialog inert-ness and CSS inert-ness in practice.

I think I need more help understanding the problematic cases here. Basically, there seem to be two types of interaction implementations: one in which modal `<dialog>` elements are used, and another where they aren’t. In the former case, those modal dialogs will always do the right thing - they’ll be non-inert while the rest of the page is inert. In the latter case, it’ll be up to the site to properly inert and un-inert the page and the “dialog”, regardless of the mechanism used for “inert”. If it’s the platform, they’ll have to get their `interactivity` properties set up correctly. If it’s totally user-space, they’ll have to make sure their JS code that handles focus trapping is done correctly. There is a potential for footguns in either approach. But I don’t think it helps to *always require* userspace modal implementations to also build their own focus trapping. That would seem to create two footguns rather than one, wouldn’t it? 


> There's also the issue that exempting modal dialogs from being able to be escaped does away with the most common use cases mentioned so far. Going back to @scottaohara's [examples](https://github.com/w3c/csswg-drafts/issues/10711#issuecomment-2379610168), two of the three would require being able to escape dialogs.

I agree, use cases 2 and 3 there would not be able to use a native modal `<dialog>` element, and would instead need to use the `interactivity` property to inert and de-inert the proper elements.



> So having said all that, the reason I think I'm opposed to adding any inert escape via CSS rather than having CSS also escape dialogs is:

Oh, I guess you are supportive of escaping the inertness of modal-`<dialog>`. Can you please confirm? 

> * Providing screen reader access to controls outside a modal dialog is more complicated than just putting them in the accessibility tree. Multiple screen readers either hard- or soft-restrict the cursor to an open dialog, so even if something else _were_ escaped and in the a11y tree, it still wouldn't necessarily be reachable. VoiceOver is the strictest about this I believe, and won't even expose live region changes outside of a dialog, even if they are in the a11y tree. This is essentially what I was chatting with Alice about earlier, which she referenced in [this comment](https://github.com/whatwg/html/issues/10811#issuecomment-2540307654)

Help me understand - this sounds like a reason *not* to allow escaping modal `<dialog>` inertness.

> * Having any control on the page able to escape being inert-ed by an open dialog based on style specificity just feels like opening the door to chaos. It introduces a war between different elements trying to make their inert styles win out, essentially moving the accessibility aspect of top-layer management into CSS.

I guess same question as above - this sounds like a reason *not* to allow escaping.

> TL;DR: `inert` is such a powerful tool for either helping or destroying accessibility that combining it with an authoring experience that is hard to understand and easy to misuse seems like a recipe for disaster.

This is a good thing to be concerned about. I think the documentation needs to be clear. And to me, these three points encapsulate the behavior:
1. Modal dialogs make the page inert, and the dialog contents non-inert. This behavior cannot be overridden.
2. The `inert` attribute simply sets `interactivity: inert` on the element.
3. The `interactivity` property can be used to set an element and its descendants to be inert (`interactivity: inert`) or un-inert (`interactivity: auto`).

While it’s three bullets instead of one, it still feels fairly understandable to me.

alice commented:
> I wonder if the better “solution” to that problem is to require developers to simply nest the popover within the dialog.

To clarify, does that mean you now support my [proposal](https://github.com/whatwg/html/issues/10811#issuecomment-2552562520) that popovers outside of the currently showing modal `<dialog>` should not be rendered on top of it?

I would still be fine with that, and I agree it gets around having to write more complicated code for the accessibility tree. That would still mean that use case doesn't justify the inert-escaping behaviour that's being pushed here.

> Can you help me understand your approach a bit better? Reading [your comment](https://github.com/whatwg/html/issues/10811#issuecomment-2540247720) and the ones below, it seems to just say that modal dialogs form a “blocking” top layer item, which un-inert the things above them in the stack, and inert everything below them. That still feels like magic (i.e. it doesn’t _explain_ in the sense of other developer-visible APIs) and also _requires_ the use of the dialog element and the UA-provided top layer.

It explains in the sense that it ties un-inerting to a lower-level concept, the top layer, rather than being specific to the `<dialog>` element. That would mean that other ways to add things to the top layer, including popover and any other future features which use the top layer, could easily be modified to allow inerting behaviour. Unlike the current language, it _doesn't_ require `<dialog>` necessarily.

What it does do, unlike this proposal, is inherently tie "everything-else-inerting" to a visual indication that content is non-interactive. Content in the top layer, particularly a top layer with a backdrop style, visually obscures content in lower layers.

> it doesn’t allow userspace modals that aren’t in the top layer.

I'm fine with that. Obviously, any page-level modal should be in the top layer anyway, so that it doesn't [conflict with other top layer elements](https://adrianroselli.com/2023/05/brief-note-on-popovers-with-dialogs.html). 

"Component modals" seem to be doing a lot of heavy lifting at this point as a justification for this API. Is that really a user need so urgent we can't, as Scott suggested, take a moment to look at alternatives? Are we really that sure, even, that this API is a good solution for those? Were any alternatives explored at all for that problem?

(Oh, sorry, UI tours too. Same comments apply.)

> Technically, this PR doesn’t add un-inerting behavior (that behavior is in the [CSS spec](https://drafts.csswg.org/css-ui/#inertness)). On the contrary, this PR goes to some length to preserve the **existing** behavior wherein modal dialog related inertness **cannot** be un-inerted.

It does add un-inerting behaviour for the `inert` attribute, though. Besides, the CSS PR already landed, so it seems unlikely any comments there would be addressed.

> This is a good thing to be concerned about. I think the documentation needs to be clear. 

I think you have a lot more faith in developers to read and absorb accessibility guidance than I do. The unfortunate reality is that a significant number of developers won't consider accessibility concerns at all, much less take the time to read the documentation. If it looks good, that's what ships. We can't fix that problem across the board, but we can at least do our best to avoid footguns that are almost _certain_ to harm users.

Scott and Sarah and I are _all_ advocating not rushing into this API, and you've acknowledged at least some of these concerns are valid. Please don't misconstrue this as a simple request for additional documentation.

lukewarlow commented:
https://jsfiddle.net/056qswnh/ - take this demo for example (relies on chrome 135), clicking on the popover above the dialog's button triggers that button, despite a popover visually being on top. Clicking outside the popover closes the popover but not the *light dismissable* dialog, despite the fact all other pointer events get swallowed down to the dialog?

I just don't know how to explain this behaviour. Using interactivity also doesn't solve this problem nor explaining of the behaviour. Meanwhile if I have a custom modal dialog I can easily set interactivity: auto on the popover and it just work. Are we just going to prolong the situation where people use custom modal implementations rather than native dialog?

I agree that if a popover is opened in the top layer above the dialog but is inert it shouldn't render. Else you end up with a very confusing user experience.

> I wonder if the better “solution” to that problem is to require developers to simply nest the popover within the dialog. That provides out-of-the-box inertness and a11y.

One other possibility of an escape hatch would be based on the popovers invoker DOM location. IIRC the location of the popover doesn't really matter for anything else in the popover API (e.g. nesting behaviour). If the invoker is a dialog child perhaps that's enough of a signal that actually that popover should work correctly even though it's outside the dialog in the DOM?

For the cases like toasts where there isn't a trigger element, you could still do something like `.showPopover({source: dialog})` to get that escape hatch?

Now I'm not saying allowing people to uninert modal dialogs whenever they want is a good thing. But if they can do common things for their custom dialog (or by using a non-modal dialog that they then hijack into a "modal") but not with the built-in it feels like another reason people would not use the built-in. That's a shame given all the hardwork that's gone into making real dialogs great (command/commandfor, close watchers, optional light dismiss etc).

If we could build these escape hatches into the system it would be good. Now I realise some of this is tangential to the point of this PR but the new author control does introduce a new angle to look at this.

smhigley commented:
Some less quick responses :D:

To @mfreed7's questions:

> Oh, I guess you are supportive of escaping the inertness of modal-<dialog>. 

and

> Help me understand - this sounds like a reason not to allow escaping modal <dialog> inertness.

More directly, I do not think escaping inertness is a good idea for the reasons Alice and Scott outlined more eloquently.

I _do_ think there are some valid use cases to address that involve components outside modals needing to be user-accessible (e.g. a toast notification center). However, I don't think a direct inert escape hatch is the best approach; instead, I think it needs more brainstorming.

One of the things I was trying to get at earlier in a roundabout way is that the primary use case for having `interactivity: auto` escape inertness is specifically for controls that need to escape modal dialogs. Since it isn't a good idea to support that, I don't really see a compelling reason to have this at all, especially since it comes with significant drawbacks and foot-guns.

mfreed7 commented:
Per the comments above, and the [recent resolution](https://github.com/w3c/csswg-drafts/issues/10711#issuecomment-2884252218) from the CSSWG/WHATWG/OpenUI joint meeting, I've updated this PR to disallow de-inerting. The only thing that remains de-inertable is the topmost modal dialog.

annevk commented on source:
```diff
@@ -80217,56 +80218,89 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
-  <p>A node (in particular elements and text nodes) can be <dfn>inert</dfn>. When a node is
-  <span>inert</span>:</p>
-
+  <p>An element or <code>Text</code> <var>node</var> must be considered <span>inert</span> if one of
+  the following is true:</p>
```
This doesn't sense to me. Are you attempting to change the CSS definition of inert here? But presumably the CSS definition already defines how it integrates with the `interactivity` property so this is at least partially redundant?

mfreed7 replied:
Well, the [CSS definition of "inert"](https://drafts.csswg.org/css-ui-4/#inert) describes what it means for something to be inert (e.g. act as if pointer-events is none). But perhaps you're talking about the [CSS definition of `interactivity`](https://drafts.csswg.org/css-ui-4/#propdef-interactivity). CSS does define that, but it leaves the behavior to the host language: "The element’s inertness is determined by the host language. (This will usually mean the element is not inert.)". So the intention of this section was to provide the ways HTML makes things "inert", namely the behavior in CSS's parenthetical.

mfreed7 commented on source:
```diff
@@ -80217,56 +80218,89 @@ interface <dfn interface>VisibilityStateEntry</dfn> : <span>PerformanceEntry</sp
+  <p>All <span>HTML elements</span> may have the <code data-x="attr-inert">inert</code>
+  attribute, a <span>boolean attribute</span>.</p>
+
+  <p>When an element has the <span>inert</span> attribute, it indicates that the element and its
```
Yep, good catch. Fixed.

annevk replied:
Why is `span` still here?
